docs/hugo/content/guide/expressions.md

```diff
@@ -0,0 +1,168 @@
+There's nothing special about Redis it's just used as an example.
+
+In each of these examples, we start with the given Redis and apply the expression as mentioned in the example itself.
+
+It might be useful to refer to 
+the [DestinationExpression](https://pkg.go.dev/github.com/Azure/azure-service-operator/v2/pkg/genruntime/core#DestinationExpression)
+documentation to understand what the `name`/`key`/`value` fields are doing in all of these examples.
+
+For more advanced examples, refer to the 
+[CEL language definition](https://github.com/google/cel-spec/blob/master/doc/langdef.md).
+
+**Sample Redis:**
+```yaml
+apiVersion: cache.azure.com/v1api20230801
+kind: Redis
+metadata:
+  name: sampleredis1
+  namespace: default
+  annotations:
+    foo: bar
+    baz: qux
+spec:
+  location: westus2
+  owner:
+    name: aso-sample-rg
+  sku:
+    family: P
+    name: Premium
+    capacity: 1
+  enableNonSslPort: false
+  minimumTlsVersion: "1.2"
+  redisConfiguration:
+    maxmemory-delta: "10"
+    maxmemory-policy: allkeys-lru
+  redisVersion: "6"
+  zones:
+  - "1"
+  - "2"
+  - "3" 
+```
+
+{{% alert title="Note" %}}
+Just to keep it simple, the sample YAML doesn't include status. Real resources have status, and you can build your
+CEL expressions with `self.status` in addition to `self.spec` or `self.metadata`.
+{{% /alert %}}
+
+#### A full example
+
+Expression snippet:
+```yaml
+spec:
+  operatorSpec:
+    configMapExpressions:
+    - name: my-configmap
+      key: location
+      value: self.spec.location
+```
+
+Output (ConfigMap):
+```yaml
+metadata:
+  name: my-configmap
+data:
+  location: westus2
+```
+
+From here on, we'll elide the supporting YAML of the above example and just focus on the CEL expression input (`value`)
+and the resulting output, still based on the example `Redis` above.
+
+| Description        | Expression                                             | Result                                                                                                  |
+|--------------------|--------------------------------------------------------|---------------------------------------------------------------------------------------------------------|
+| Hardcoded string   | `"helloworld"`                                         | `helloworld`                                                                                            |
+| Formatted string   | `"%s:%d".format([self.spec.location, 7])`              | `westus2:7`                                                                                             |
+| String math        | `self.metadata.namespace + ":" + self.metadata.name`   | `default:sampleredis1`                                                                                  |
+| Int output (error) | `self.spec.sku.capacity`                               | Error, expression "self.spec.sku.capacity" must return one of [string,map(string, string)], but was int |
```

Comment:
We technically could, but it would require us to coerce it in Go, not in CEL. What we could do is support more than just `string` and `map[string]string` outputs and do the coercion ourselves afterwards, effectively expanding the set of allowed types from `string` and `map[string]string` to include others.

The problem with that is: 
- What types do we covert? Everything or just ints? Just ints feels incomplete, but if we start converting other types like doubles too it gets more complicated as you then have to talk about precision and how many decimal points you keep in the string representation - different users may want different values, etc.
- Do we convert types in _all_ contexts or only in the "single-return" context. What I mean is, say we auto-converted ints. Do we also auto-convert map keys that are ints to be strings? What about map values?

I don't know in practice how often users are going to export ints. If you look at all of the existing fields that we export via ConfigMap every one of them seems to natively be a string. For secrets there are a few ints (Redis `Port` and `SSLPort` for example), but at least based on existing exports, pure strings are the 95% case.

All the above suggests to me that the best thing to do for user clarity is to keep it simple: The expression output type must be `string` or `map[string]string`, and if you're dealing with types _other_ than those it's up to you to convert them yourself using inbuilt CEL helpers to do so. This keeps our life easy, keeps it easy to explain how expressions work to users, and gives users the most control to do exactly what they want without relying on magic on our end.

Reply:
Fair enough.